# 2 - PREPARE

In this section I will import the monthly datasets and concatenate them to create a single DataFrame called 'df'.

In [31]:
#import packages
import pandas as pd
import os

In [32]:
#importing 1 csv at a time
desktop_path1 = "/Users/amylee/Desktop/raw_data/202401-divvy-tripdata.csv"
df1 = pd.read_csv(desktop_path1)

In [33]:
#show first 5 rows
df1.head()

ride_id  rideable_type           started_at             ended_at  \
0  C1D650626C8C899A  electric_bike  2024-01-12 15:30:27  2024-01-12 15:37:59   
1  EECD38BDB25BFCB0  electric_bike  2024-01-08 15:45:46  2024-01-08 15:52:59   
2  F4A9CE78061F17F7  electric_bike  2024-01-27 12:27:19  2024-01-27 12:35:19   
3  0A0D9E15EE50B171   classic_bike  2024-01-29 16:26:17  2024-01-29 16:56:06   
4  33FFC9805E3EFF9A   classic_bike  2024-01-31 05:43:23  2024-01-31 06:09:35   

           start_station_name start_station_id           end_station_name  \
0           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
1           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
2           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
3      Wells St & Randolph St     TA1305000030  Larrabee St & Webster Ave   
4  Lincoln Ave & Waveland Ave            13253   Kingsbury St & Kinzie St   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0   KA1503000043  41.903267 -87.634737  41.889177 -87.638506        member  
1   KA1503000043  41.902937 -87.634440  41.889177 -87.638506        member  
2   KA1503000043  41.902951 -87.634470  41.889177 -87.638506        member  
3          13193  41.884295 -87.633963  41.921822 -87.644140        member  
4   KA1503000043  41.948797 -87.675278  41.889177 -87.638506        member

In [34]:
import pandas as pd
import os

# Define the base path to your data directory
base_path = "/Users/amylee/Desktop/raw_data/" 

# Create an empty dictionary to store the DataFrames
dataframes = {}

# Iterate through months 1 to 11
for month in range(1, 12):
    # Construct the filename for each month
    filename = f"2024{month:02d}-divvy-tripdata.csv"  # Pad month with leading zero (e.g., '01' for Jan)
    filepath = os.path.join(base_path, filename)

    # Read csv file into a dataframe
    try:
        df = pd.read_csv(filepath)
        dataframes[f"df{month}"] = df  # Store the DataFrame in the dictionary
    except FileNotFoundError:
        print(f"Warning: File not found: {filepath}")

In [35]:
# Now you have DataFrames named df1, df2, ..., df11 in the 'dataframes' dictionary
print(dataframes["df1"].head())  # Example: Access and display the first DataFrame

            ride_id  rideable_type           started_at             ended_at  \
0  C1D650626C8C899A  electric_bike  2024-01-12 15:30:27  2024-01-12 15:37:59   
1  EECD38BDB25BFCB0  electric_bike  2024-01-08 15:45:46  2024-01-08 15:52:59   
2  F4A9CE78061F17F7  electric_bike  2024-01-27 12:27:19  2024-01-27 12:35:19   
3  0A0D9E15EE50B171   classic_bike  2024-01-29 16:26:17  2024-01-29 16:56:06   
4  33FFC9805E3EFF9A   classic_bike  2024-01-31 05:43:23  2024-01-31 06:09:35   

           start_station_name start_station_id           end_station_name  \
0           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
1           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
2           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
3      Wells St & Randolph St     TA1305000030  Larrabee St & Webster Ave   
4  Lincoln Ave & Waveland Ave            13253   Kingsbury St & Kinzie St   

  end_station_id  start_lat  start_lng    end_lat    end

In [36]:
#extract the DataFrames from the dictionary
df1 = dataframes['df1']
df2 = dataframes['df2']
df3 = dataframes['df3']
df4 = dataframes['df4']
df5 = dataframes['df5']
df6 = dataframes['df6']
df7 = dataframes['df7']
df8 = dataframes['df8']
df9 = dataframes['df9']
df10 = dataframes['df10']
df11 = dataframes['df11']

In [37]:
#create a list of all dataframes
dataframes_list = [df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11]

In [38]:
#concatenate all DataFrames into a single DataFrame
df = pd.concat(dataframes_list, ignore_index=True)

In [39]:
#print the first few rows of the combined DataFrame
df.head()

ride_id  rideable_type           started_at             ended_at  \
0  C1D650626C8C899A  electric_bike  2024-01-12 15:30:27  2024-01-12 15:37:59   
1  EECD38BDB25BFCB0  electric_bike  2024-01-08 15:45:46  2024-01-08 15:52:59   
2  F4A9CE78061F17F7  electric_bike  2024-01-27 12:27:19  2024-01-27 12:35:19   
3  0A0D9E15EE50B171   classic_bike  2024-01-29 16:26:17  2024-01-29 16:56:06   
4  33FFC9805E3EFF9A   classic_bike  2024-01-31 05:43:23  2024-01-31 06:09:35   

           start_station_name start_station_id           end_station_name  \
0           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
1           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
2           Wells St & Elm St     KA1504000135   Kingsbury St & Kinzie St   
3      Wells St & Randolph St     TA1305000030  Larrabee St & Webster Ave   
4  Lincoln Ave & Waveland Ave            13253   Kingsbury St & Kinzie St   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0   KA1503000043  41.903267 -87.634737  41.889177 -87.638506        member  
1   KA1503000043  41.902937 -87.634440  41.889177 -87.638506        member  
2   KA1503000043  41.902951 -87.634470  41.889177 -87.638506        member  
3          13193  41.884295 -87.633963  41.921822 -87.644140        member  
4   KA1503000043  41.948797 -87.675278  41.889177 -87.638506        member

In [40]:
#high level view of dataframe info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5682196 entries, 0 to 5682195
Data columns (total 13 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
dtypes: float64(4), object(9)
memory usage: 563.6+ MB


### PREPARE SUMMARY

In this section I uploaded the monthly csv data to my working directory. I read the csv files in as individual dataframes then combined them into 1 combined dataframe called 'df'.

The dataframe has 12 columns with 5,682,196 entries.

# PROCESS

In this section I will complete the following key tasks:
     1.      Check the data for errors.
     2.      Transform the data so I can work with it effectively.
     3.      Document the cleaning process.

In [41]:
df['started_at'] = pd.to_datetime(df['started_at'], format='%Y-%m-%d %H:%M:%S')

ValueError: unconverted data remains when parsing with format "%Y-%m-%d %H:%M:%S": ".289", at position 1694242. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

In [42]:
#investigating problematic row - started_at does not conform to datetime format
problem_row = df.loc[1694242]
print(problem_row)

ride_id                      CDE6023BE6B11D2F
rideable_type                   electric_bike
started_at            2024-06-11 17:20:06.289
ended_at              2024-06-11 17:21:39.464
start_station_name                        NaN
start_station_id                          NaN
end_station_name                          NaN
end_station_id                            NaN
start_lat                               41.89
start_lng                              -87.65
end_lat                                 41.89
end_lng                                -87.65
member_casual                          casual
Name: 1694242, dtype: object


In [43]:
#count observations with fractional seconds in 'started_at'
count_frxnal_seconds = df['started_at'].str.contains(r'\.\d+').sum()
print(f"Number of observations with fractional seconds in 'started_at' column: {count_frxnal_seconds}")

Number of observations with fractional seconds in 'started_at' column: 3987954


In [44]:
#ratio of dataset that has fractional seconds
ratio_frxnal_seconds = count_frxnal_seconds/(len(df))*100
print(f"Percent of dataset with fractional seconds: {ratio_frxnal_seconds}")

Percent of dataset with fractional seconds: 70.18332348972123


70% of the dataset has fractional seconds. Therefore, I cannot easily convert 'started_at' and 'ended_at' columns to datetime format. I will 
add .000 to the end of the 30% of the 'started_at'/'ended_at' that did not have fractional seconds and round. 

***For the purposes of this analysis this type of precision is not necessary. Removing the fractional seconds then converting to datetime would have sufficed.

In [45]:
def add_fractional_seconds(df):
    """
    Adds '.000' to timestamps in 'started_at' and 'ended_at' columns that do not alread have fractional seconds.
    
    Args:
        df: The pandas DataFrame
        
    Returns:
        The DataFrame with modified timestamps.
    """
    
    df['started_at'] = df['started_at'].apply(lambda x: x+'.000' if not '.' in x else x)
    df['ended_at'] = df['ended_at'].apply(lambda x: x+'.000' if not '.' in x else x)
    return df

#re-establish df
df = add_fractional_seconds(df)

In [46]:
#convert 'started_at' and 'ended_at' columns to datetime
df['started_at'] = pd.to_datetime(df['started_at'])
df['ended_at'] = pd.to_datetime(df['ended_at'])

In [47]:
#calculate and create a column called 'ride_length'
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5682196 entries, 0 to 5682195
Data columns (total 13 columns):
 #   Column              Dtype         
---  ------              -----         
 0   ride_id             object        
 1   rideable_type       object        
 2   started_at          datetime64[ns]
 3   ended_at            datetime64[ns]
 4   start_station_name  object        
 5   start_station_id    object        
 6   end_station_name    object        
 7   end_station_id      object        
 8   start_lat           float64       
 9   start_lng           float64       
 10  end_lat             float64       
 11  end_lng             float64       
 12  member_casual       object        
dtypes: datetime64[ns](2), float64(4), object(7)
memory usage: 563.6+ MB


In [48]:
# count number of null values in each column
null_counts = df.isnull().sum()

# calculate the % of null values in each column
null_percents = (null_counts/len(df)) * 100

#create a DataFrame to display results
null_summary = pd.DataFrame({'Null Count': null_counts,
                            'Null Percentage': null_percents.round(2)})

# print results
print("Null Value Summary:\n", null_summary)

Null Value Summary:
                     Null Count  Null Percentage
ride_id                      0             0.00
rideable_type                0             0.00
started_at                   0             0.00
ended_at                     0             0.00
start_station_name     1044760            18.39
start_station_id       1044760            18.39
end_station_name       1073877            18.90
end_station_id         1073877            18.90
start_lat                    0             0.00
start_lng                    0             0.00
end_lat                   7101             0.12
end_lng                   7101             0.12
member_casual                0             0.00


In [49]:
# drop rows containing null values in end_lat and end_lng columns

df = df.dropna(subset=['end_lat', 'end_lng'])
print(df.isnull().sum())

ride_id                     0
rideable_type               0
started_at                  0
ended_at                    0
start_station_name    1044760
start_station_id      1044760
end_station_name      1066776
end_station_id        1066776
start_lat                   0
start_lng                   0
end_lat                     0
end_lng                     0
member_casual               0
dtype: int64


In [50]:
null_counts.info()

<class 'pandas.core.series.Series'>
Index: 13 entries, ride_id to member_casual
Series name: None
Non-Null Count  Dtype
--------------  -----
13 non-null     int64
dtypes: int64(1)
memory usage: 764.0+ bytes


In [51]:
def count_decimal_places(df, col_name):
  """
  Counts the number of observations with each amount of decimal places 
  in the specified column of a DataFrame.

  Args:
    df: The pandas DataFrame.
    col_name: The name of the column to analyze.

  Returns:
    A pandas Series containing the counts of observations for each 
    number of decimal places.
  """

  # Extract the decimal part of the column
  decimal_parts = df['start_lat'].astype(str).str.split('.').str[1] 

  # Handle cases where there are no decimal places
  decimal_parts = decimal_parts.fillna('') 

  # Count the number of decimal places for each observation
  decimal_place_counts = decimal_parts.str.len() 

  # Count the occurrences of each decimal place count
  decimal_place_counts = decimal_place_counts.value_counts()

  return decimal_place_counts

In [52]:
import pandas as pd

def count_decimal_places(df, col_name):
  """
  Counts the number of observations with each amount of decimal places 
  in the specified column of a DataFrame.

  Args:
    df: The pandas DataFrame.
    col_name: The name of the column to analyze.

  Returns:
    A pandas Series containing the counts of observations for each 
    number of decimal places.
  """

  # Extract the decimal part of the column
  decimal_parts = df[col_name].astype(str).str.split('.').str[1] 

  # Handle cases where there are no decimal places
  decimal_parts = decimal_parts.fillna('') 

  # Count the number of decimal places for each observation
  decimal_place_counts = decimal_parts.str.len() 

  # Count the occurrences of each decimal place count
  decimal_place_counts = decimal_place_counts.value_counts()

  return decimal_place_counts

# Assuming 'df' is your DataFrame
start_lat_decimal_counts = count_decimal_places(df, 'start_lat')
end_lat_decimal_counts = count_decimal_places(df, 'end_lat')

start_lng_decimal_counts = count_decimal_places(df, 'start_lng')
end_lng_decimal_counts = count_decimal_places(df, 'end_lng')


print("Decimal Place Counts for 'start_lat':\n", start_lat_decimal_counts)
print("\nDecimal Place Counts for 'end_lat':\n", end_lat_decimal_counts)

print("Decimal Place Counts for 'start_lng':\n", start_lng_decimal_counts)
print("\nDecimal Place Counts for 'end_lng':\n", end_lng_decimal_counts)

Decimal Place Counts for 'start_lat':
 start_lat
6     2126737
2      936765
9      741556
14     540994
11     289371
5      257011
15     163179
8      132154
10     122541
7      122479
1      107999
13      91300
4       42485
12        364
3         160
Name: count, dtype: int64

Decimal Place Counts for 'end_lat':
 end_lat
6     2635945
2      956626
14     572520
11     360895
5      332355
15     163199
10     152161
13     110333
1      110150
7      102100
8       72748
4       52918
9       52396
12        496
3         253
Name: count, dtype: int64
Decimal Place Counts for 'start_lng':
 start_lng
6     2007678
2      998002
9      725207
14     724408
10     423322
5      323268
8      161715
7      108763
13      74719
4       69639
1       46760
3       11614
Name: count, dtype: int64

Decimal Place Counts for 'end_lng':
 end_lng
6     2492149
2     1018481
14     753292
10     523836
5      411040
8      107357
13      93262
4       90419
7       83584
1       48295
9   

In [53]:
# decided that for this analysis lat/lng data is sufficient, station name/id is not needed. Therefore I will drop 4 colummns: start_station_name, start_station_id, end_station_name, end_station_id

df = df.drop(columns=['start_station_name', 'start_station_id', 'end_station_name', 'end_station_id'])

In [54]:
print(df.isnull().sum())

ride_id          0
rideable_type    0
started_at       0
ended_at         0
start_lat        0
start_lng        0
end_lat          0
end_lng          0
member_casual    0
dtype: int64


In [55]:
# check for duplicates

#find duplicate rows based on all columns
duplicate_rows = df[df.duplicated()]

In [56]:
# count the number of duplicate rows
num_duplicates = len(duplicate_rows)

# calculate percentage duplicates make up of dataset 
percent_duplicates = num_duplicates / len(df) * 100

print(f"Number of duplicate rows: {num_duplicates}")
print(f"% of dataset that are duplicates: {percent_duplicates}")

Number of duplicate rows: 0
% of dataset that are duplicates: 0.0


In [57]:
# previously had double imported august data. This was part of trouble shooting. Fixed the original import, this section no longer relevant.
# duplicate_rows.groupby(['ride_id'])
# print(duplicate_rows.groupby(['ride_id']).head(20))

In [58]:
# error fixed - cell not needed
# check for duplicates

# find duplicate rows based on ride_id column
# duplicate_ride_ids = df['ride_id'].duplicated(keep=False)

In [59]:
# error fixed - cell not needed 
# filter the DataFrame to include only rows with duplicate 'ride_id'
# rows_with_duplicate_ride_ids = df[duplicate_ride_ids]

In [60]:
# error fixed - cell not needed 
# sort the duplicate rows by 'ride_id'
# rows_with_duplicate_ride_ids = rows_with_duplicate_ride_ids.sort_values(by='ride_id')

In [61]:
# print the sorted duplicate rows
# print(rows_with_duplicate_ride_ids)

In [62]:
# rows_with_duplicate_ride_ids.info()

In [63]:
# rows_with_duplicate_ride_ids.groupby(['ride_id'])
# print(rows_with_duplicate_ride_ids.head())

In [64]:
# rows_with_duplicate_ride_ids.groupby(['ride_id'])
# print(rows_with_duplicate_ride_ids.tail())

In [65]:
# example_duplicate = df[df['ride_id'] == '00001004F8DF01C8']

In [66]:
# print the selected rows
# print(example_duplicate)

All duplicates seem to be in the month of August. Checking imported dfs to make sure August was not somehow uploaded twice.

In [67]:
# error fixed - cell not needed 
# df8.info()
# df8.head()

In [68]:
# error fixed - cell not needed 
# print(len(df8))

columns 10 (end_lat) and 11 (end_lng) have 754612 non-null values --> the same as number of duplicates listed. Must be a code error.

Error identified in cell 20: original import brought in august twice and no september. Fixed.

In [69]:
# calculate ride length and add new column to dataframe

df['ride_length'] = (df['ended_at'] - df['started_at'])
print(df.head())

            ride_id  rideable_type          started_at            ended_at  \
0  C1D650626C8C899A  electric_bike 2024-01-12 15:30:27 2024-01-12 15:37:59   
1  EECD38BDB25BFCB0  electric_bike 2024-01-08 15:45:46 2024-01-08 15:52:59   
2  F4A9CE78061F17F7  electric_bike 2024-01-27 12:27:19 2024-01-27 12:35:19   
3  0A0D9E15EE50B171   classic_bike 2024-01-29 16:26:17 2024-01-29 16:56:06   
4  33FFC9805E3EFF9A   classic_bike 2024-01-31 05:43:23 2024-01-31 06:09:35   

   start_lat  start_lng    end_lat    end_lng member_casual     ride_length  
0  41.903267 -87.634737  41.889177 -87.638506        member 0 days 00:07:32  
1  41.902937 -87.634440  41.889177 -87.638506        member 0 days 00:07:13  
2  41.902951 -87.634470  41.889177 -87.638506        member 0 days 00:08:00  
3  41.884295 -87.633963  41.921822 -87.644140        member 0 days 00:29:49  
4  41.948797 -87.675278  41.889177 -87.638506        member 0 days 00:26:12  


In [70]:
# calculate day_of_week and add column to dataframe
df['day_of_week'] = df['started_at'].dt.day_name()
df.head()

ride_id  rideable_type          started_at            ended_at  \
0  C1D650626C8C899A  electric_bike 2024-01-12 15:30:27 2024-01-12 15:37:59   
1  EECD38BDB25BFCB0  electric_bike 2024-01-08 15:45:46 2024-01-08 15:52:59   
2  F4A9CE78061F17F7  electric_bike 2024-01-27 12:27:19 2024-01-27 12:35:19   
3  0A0D9E15EE50B171   classic_bike 2024-01-29 16:26:17 2024-01-29 16:56:06   
4  33FFC9805E3EFF9A   classic_bike 2024-01-31 05:43:23 2024-01-31 06:09:35   

   start_lat  start_lng    end_lat    end_lng member_casual     ride_length  \
0  41.903267 -87.634737  41.889177 -87.638506        member 0 days 00:07:32   
1  41.902937 -87.634440  41.889177 -87.638506        member 0 days 00:07:13   
2  41.902951 -87.634470  41.889177 -87.638506        member 0 days 00:08:00   
3  41.884295 -87.633963  41.921822 -87.644140        member 0 days 00:29:49   
4  41.948797 -87.675278  41.889177 -87.638506        member 0 days 00:26:12   

  day_of_week  
0      Friday  
1      Monday  
2    Saturday  
3      Monday  
4   Wednesday

#### Save the data

In [71]:
df.shape

(5675095, 11)

Save this to your data directory, separately. Uploaded raw data as monthly csvs. Should save derived data in a separate location. This guards against overwriting our original data.

In [72]:
def save_file(df, filename, path):
    """
    Saves a pandas DataFrame to a CSV file.

    Args:
        df: The pandas DataFrame to save.
        filename: The name of the CSV file.
        path: The path to the directory where the file should be saved.
    """
    filepath = f"{path}/{filename}" 
    df.to_csv(filepath, index=False)

In [73]:
# save the data to a new csv file
datapath = '/Users/amylee/google_data_analytics/cyclistic_case_study/cyclistic_data'
save_file(df, 'df_cyclistic_clean.csv', datapath)

In [74]:
%pwd

'/Users/amylee/google_data_analytics/cyclistic_case_study'

### PROCESS SUMMARY

Data cleaning tasks completed in this section:
1. Ensured all columns were in appropriate data format - converted 'started_at' and 'ended_at' to datetime
2. Addressed/eliminated null values. 
a - Dropped rows with null values in 'end_lat' and 'end_lng' columns (~0.01% of dataset).
b - Dropped 4 columns: 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id'. 18% of observations in these columns were missing data. Dataset also contains start/end latitude/longitude so that will be used for geographical analysis. 
5. Checked for duplicates - none. Identified import error and corrected.

I then added columns to calculate ride_length and the day_of_week for each ride.